### ISSUE TITLE
Insertion and Mutation of different Components on the same `EntityMut`

### ISSUE THREAD

**Issue by Zeenobit** [bevyengine/bevy#4155]
## Bevy version

0.6

## Operating system & version

Windows 10 64-bit

## What you did

The following assertion fails if I uncomment the line mentioned in the snippet below:

```rust
fn what() {
    #[derive(Component)]
    struct Dummy(u64);

    #[derive(Component)]
    struct OtherDummy;

    let mut world = World::default();
    let entity: Entity;

    {
        entity = world.spawn().insert(Dummy(0)).id();
    }

    {
        let mut entity = world.entity_mut(entity);
        let mut dummy = entity.get_mut::<Dummy>().unwrap();
        //entity.insert(OtherDummy); <--- Uncommenting this line makes the test fail?!
        dummy.0 = 12;
    }

    {
        let entity = world.entity(entity);
        let dummy = entity.get::<Dummy>().unwrap();

        assert_eq!(dummy.0, 12);
    }
}
```

## What you expected to happen

Insertion or removal of an unrelated component shouldn't change the result of the test.

## What actually happened

The test doesn't fails because value of `dummy` is not set correctly.

## Additional information

This issue was raised from a discussion on Discord:
https://discord.com/channels/691052431525675048/950812276712165416

As a workaround, re-ordering the insertion/mutation steps seems to pass the test:
```rust
fn what() {
    #[derive(Component)]
    struct Dummy(u64);

    #[derive(Component)]
    struct OtherDummy;

    let mut world = World::default();
    let entity: Entity;

    {
        entity = world.spawn().insert(Dummy(0)).id();
    }

    {
        let mut entity = world.entity_mut(entity);
        entity.insert(OtherDummy);
        let mut dummy = entity.get_mut::<Dummy>().unwrap();
        dummy.0 = 12;
    }

    {
        let entity = world.entity(entity);
        let dummy = entity.get::<Dummy>().unwrap();

        assert_eq!(dummy.0, 12);
    }
}
```

**Comment by DJMcNab** [comment]
Duplicate of https://github.com/bevyengine/bevy/issues/3408

**Comment by hymm** [comment]
Can confirm this is fixed by #3001.  Should close this out in favor of the other issue.
